Error when differentiating a non-void function with its return unused in the reverse mode

Currently, if a function call of non-void return type has no dfdx, we have a crash. An easy reproducer:
```
double g(double& x) {
  return x++;
}
  
double f(double x, double y) {
  g(x);
  return x;
}

int main() {
  clad::gradient(f);
  return 0;
}
```
produces
```
void g_pullback(double &x, double _d_y, double *_d_x) {
    *_d_x += _d_y;
}
clangtest.cpp:14:5: error: too few arguments to function call, expected 3, have 2
    g(x);
    ^
note: 'g_pullback' declared here
void f_grad(double x, double y, double *_d_x, double *_d_y) {
    double _t0 = x;
    g(x);
    *_d_x += 1;
    x = _t0;
}
1 error generated.
```

--- a/include/clad/Differentiator/BuiltinDerivatives.h
+++ b/include/clad/Differentiator/BuiltinDerivatives.h
@@ -89,8 +89,8 @@ __global__ void atomicAdd_kernel(T* destPtr, T* srcPtr, size_t N) {
 
 template <typename T>
 void cudaMemcpy_pullback(T* destPtr, const T* srcPtr, size_t count,
-                         cudaMemcpyKind kind, T* d_destPtr, T* d_srcPtr,
-                         size_t* d_count, cudaMemcpyKind* d_kind)
+                         cudaMemcpyKind kind, cudaError_t d_y, T* d_destPtr,
+                         T* d_srcPtr, size_t* d_count, cudaMemcpyKind* d_kind)
     __attribute__((host)) {
   T* aux_destPtr = nullptr;
   if (kind == cudaMemcpyDeviceToHost) {

--- a/lib/Differentiator/ReverseModeVisitor.cpp
+++ b/lib/Differentiator/ReverseModeVisitor.cpp
@@ -1808,14 +1808,15 @@ Expr* ReverseModeVisitor::getStdInitListSizeExpr(const Expr* E) {
       }
       Expr* pullback = dfdx();
 
-      if ((pullback == nullptr) && FD->getReturnType()->isLValueReferenceType())
-        pullback = getZeroInit(FD->getReturnType().getNonReferenceType());
-
-      if (FD->getReturnType()->isVoidType()) {
-        assert(pullback == nullptr && FD->getReturnType()->isVoidType() &&
+      if (returnType->isVoidType()) {
+        assert(pullback == nullptr && returnType->isVoidType() &&
                "Call to function returning void type should not have any "
                "corresponding dfdx().");
       }
+
+      if ((pullback == nullptr) &&
+          !(returnType->isPointerType() || returnType->isVoidType()))
+        pullback = getZeroInit(returnType.getNonReferenceType());
 
       for (Expr* arg : DerivedCallOutputArgs)
         if (arg)

--- a/test/CUDA/GradientKernels.cu
+++ b/test/CUDA/GradientKernels.cu
@@ -490,13 +490,13 @@ double fn_memory(double *out, double *in) {
 //CHECK-NEXT:    {
 //CHECK-NEXT:        unsigned long _r2 = 0UL;
 //CHECK-NEXT:        cudaMemcpyKind _r3 = static_cast<cudaMemcpyKind>(0U);
-//CHECK-NEXT:        clad::custom_derivatives::cudaMemcpy_pullback(out_host, out, 10 * sizeof(double), cudaMemcpyDeviceToHost, _d_out_host, _d_out, &_r2, &_r3);
+//CHECK-NEXT:        clad::custom_derivatives::cudaMemcpy_pullback(out_host, out, 10 * sizeof(double), cudaMemcpyDeviceToHost, static_cast<cudaError>(0U), _d_out_host, _d_out, &_r2, &_r3);
 //CHECK-NEXT:    }
 //CHECK-NEXT:    kernel_call_pullback<<<1, 10>>>(out, in_dev, _d_out, _d_in_dev);
 //CHECK-NEXT:    {
 //CHECK-NEXT:        unsigned long _r0 = 0UL;
 //CHECK-NEXT:        cudaMemcpyKind _r1 = static_cast<cudaMemcpyKind>(0U);
-//CHECK-NEXT:        clad::custom_derivatives::cudaMemcpy_pullback(in_dev, in, 10 * sizeof(double), cudaMemcpyHostToDevice, _d_in_dev, _d_in, &_r0, &_r1);
+//CHECK-NEXT:        clad::custom_derivatives::cudaMemcpy_pullback(in_dev, in, 10 * sizeof(double), cudaMemcpyHostToDevice, static_cast<cudaError>(0U), _d_in_dev, _d_in, &_r0, &_r1);
 //CHECK-NEXT:    }
 //CHECK-NEXT:    free(out_host);
 //CHECK-NEXT:    free(_d_out_host);
@@ -599,7 +599,7 @@ void launch_add_kernel_4(int *out, int *in, const int N) {
 //CHECK-NEXT:    {
 //CHECK-NEXT:        unsigned long _r6 = 0UL;
 //CHECK-NEXT:        cudaMemcpyKind _r7 = static_cast<cudaMemcpyKind>(0U);
-//CHECK-NEXT:        clad::custom_derivatives::cudaMemcpy_pullback(out, out_dev, N * sizeof(int), cudaMemcpyDeviceToHost, _d_out, _d_out_dev, &_r6, &_r7);
+//CHECK-NEXT:        clad::custom_derivatives::cudaMemcpy_pullback(out, out_dev, N * sizeof(int), cudaMemcpyDeviceToHost, static_cast<cudaError>(0U), _d_out, _d_out_dev, &_r6, &_r7);
 //CHECK-NEXT:        _d_N += _r6 * sizeof(int);
 //CHECK-NEXT:    }
 //CHECK-NEXT:    {
@@ -615,13 +615,13 @@ void launch_add_kernel_4(int *out, int *in, const int N) {
 //CHECK-NEXT:    {
 //CHECK-NEXT:        unsigned long _r2 = 0UL;
 //CHECK-NEXT:        cudaMemcpyKind _r3 = static_cast<cudaMemcpyKind>(0U);
-//CHECK-NEXT:        clad::custom_derivatives::cudaMemcpy_pullback(out_dev, out, N * sizeof(int), cudaMemcpyHostToDevice, _d_out_dev, _d_out, &_r2, &_r3);
+//CHECK-NEXT:        clad::custom_derivatives::cudaMemcpy_pullback(out_dev, out, N * sizeof(int), cudaMemcpyHostToDevice, static_cast<cudaError>(0U), _d_out_dev, _d_out, &_r2, &_r3);
 //CHECK-NEXT:        _d_N += _r2 * sizeof(int);
 //CHECK-NEXT:    }
 //CHECK-NEXT:    {
 //CHECK-NEXT:        unsigned long _r0 = 0UL;
 //CHECK-NEXT:        cudaMemcpyKind _r1 = static_cast<cudaMemcpyKind>(0U);
-//CHECK-NEXT:        clad::custom_derivatives::cudaMemcpy_pullback(in_dev, in, N * sizeof(int), cudaMemcpyHostToDevice, _d_in_dev, _d_in, &_r0, &_r1);
+//CHECK-NEXT:        clad::custom_derivatives::cudaMemcpy_pullback(in_dev, in, N * sizeof(int), cudaMemcpyHostToDevice, static_cast<cudaError>(0U), _d_in_dev, _d_in, &_r0, &_r1);
 //CHECK-NEXT:        _d_N += _r0 * sizeof(int);
 //CHECK-NEXT:    }
 //CHECK-NEXT:    cudaFree(in_dev);

--- a/test/Gradient/FunctionCalls.C
+++ b/test/Gradient/FunctionCalls.C
@@ -880,6 +880,36 @@ double fn23(double x) {
 // CHECK-NEXT:     }
 // CHECK-NEXT: }
 
+double unused_return(double& x) {
+  return x *= 2;
+}
+
+// CHECK:  void unused_return_pullback(double &x, double _d_y, double *_d_x) {
+// CHECK-NEXT:      double _t0 = x;
+// CHECK-NEXT:      {
+// CHECK-NEXT:          *_d_x += _d_y;
+// CHECK-NEXT:          x = _t0;
+// CHECK-NEXT:          double _r_d0 = *_d_x;
+// CHECK-NEXT:          *_d_x = 0.;
+// CHECK-NEXT:          *_d_x += _r_d0 * 2;
+// CHECK-NEXT:      }
+// CHECK-NEXT:  }
+
+double fn24(double x) {
+  unused_return(x);
+  return x;
+}
+
+// CHECK:  void fn24_grad(double x, double *_d_x) {
+// CHECK-NEXT:      double _t0 = x;
+// CHECK-NEXT:      unused_return(x);
+// CHECK-NEXT:      *_d_x += 1;
+// CHECK-NEXT:      {
+// CHECK-NEXT:          x = _t0;
+// CHECK-NEXT:          unused_return_pullback(_t0, 0., &*_d_x);
+// CHECK-NEXT:      }
+// CHECK-NEXT:  }
+
 template<typename T>
 void reset(T* arr, int n) {
   for (int i=0; i<n; ++i)
@@ -994,6 +1024,9 @@ int main() {
 
   INIT(fn23);
   TEST1(fn23, 3);  // CHECK-EXEC: {1.00}
+
+  INIT(fn24);
+  TEST1(fn24, 3);  // CHECK-EXEC: {2.00}
 }
 
 double sq_defined_later(double x) {

